### PR TITLE
Add an overview of what lowerings are done [DOC-25]

### PR DESCRIPTION
**Pull request by thk123**
As expressed on #2775 we should have one place that explains what is meant by lowering in a holistic sense, with links to the new sections.

### CODE REVIEW

File: jbmc/src/java_bytecode/README.md
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: high-level

Comment (peterschrammel) [reply jbmc/src/java_bytecode/README.md]: it -> Java programs?

File: jbmc/src/java_bytecode/README.md
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: Java

Comment (peterschrammel) [reply jbmc/src/java_bytecode/README.md]: `codet`

File: jbmc/src/java_bytecode/README.md
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: This is not a lowering in the strict sense. It is adding part of the JVM's execution algorithm into the analysed program.

File: jbmc/src/java_bytecode/README.md
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: There is remove_instanceof and remove_virtual_methods

File: jbmc/src/java_bytecode/README.md
Comment (peterschrammel) [reply jbmc/src/java_bytecode/README.md]: We should also mention remove_nondet then.

File: jbmc/src/java_bytecode/README.md
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: analysed: maybe be more specific 'handled by \ref goto-symex'. Also the lowering passes do some kind of 'analysis'.

File: jbmc/src/java_bytecode/README.md
Comment (thk123) [reply jbmc/src/java_bytecode/README.md]: remove-virtual-methods is not Java specific (covered by the link to goto-programs). The section for `remove_instanceof` is missing so will add after #2775 is merged.

File: jbmc/src/java_bytecode/README.md
Comment (thk123) [reply jbmc/src/java_bytecode/README.md]: Hmm - I consider it in the same bucket since the JVM detects null pointer access and throws an exception - we lower that "high level feature" of the JVM into a fundamental operation of `if(a == null)`, what would you call it? 

Remove_nondet is kind of a weird one since it doesn't exist in Java bytecode or the JVM - it is a CPROVER construct right?

**Comment by thk123** [comment]
@peterschrammel some comments addressed - I have questions about the others.

**Comment by cesaro** [comment]
@thk123 The term lowering will be defined in the section on "background concepts", subsection "glossary". There is already a sketch of the definition in https://github.com/diffblue/cbmc/pull/2806

File: jbmc/src/java_bytecode/README.md
Comment (tautschnig) [reply jbmc/src/java_bytecode/README.md]: I think we should carefully distinguish between lowering and instrumentation. The former is a rewrite of the code (supposedly semantics-preserving), while the latter will add or modify behaviour.

**Comment by thk123** [comment]
@cesaro Hi - sorry is there an action you want me to take from your comment? Perhaps a link to this section: https://github.com/diffblue/cbmc/pull/2806/files#diff-6988ca396abc0c4e424da9d8494e2f1fR814 from the word lowering. 

If I do that, would you be happy to approve?

File: jbmc/src/java_bytecode/README.md
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: Remove `new` calls

File: jbmc/src/java_bytecode/README.md
Comment (peterschrammel) [inline jbmc/src/java_bytecode/README.md]: non-Java

**Review by peterschrammel — APPROVED** [review]
Good to go after fixing the last nitpicks.

**Comment by thk123** [comment]
Done, will wait until I've merged the other (colliding) doc change to avoid spamming Travis #2773

**Comment by cesaro** [comment]
@thk123 Yes, good to merge with just a reference to the background concepts. Since this will probably be merged before #2806 you can just a reference to the section Background Concepts in plain English when you explain what a lowering are and we will create the link from here in #2806 to the specific term in the Glossary subsection.

**Pull request merged**
